Switch public APIs from File to Path

rpalcolea opened the pull request:
Related to https://github.com/Netflix/hollow/issues/167

- [x] for all methods that accept java.io.File create an overload that accepts java.nio.Path
- [x] mark java.io.File methods as @Deprecated, use file.asPath() to delegate to new methods
- [x]  use idiomatic Path and Paths code for opening files for reading/writing, interacting with directories, etc.

Please let me know your thoughts.

I was wondering if Hollow 3.x will lean towards Java NIO `Channels` and `ByteBuffer`. Guess that would add complexity to parse the data in comparison with a blocked stream as of today in things like `HollowBlob`

Also, since this is targeted for Hollow 3.0, what's the plan for branching? will 2.x get a branch for patches? or how do you envision this? wondering if this PR should go to `master`. Also I marked a the new signatures as `@since 3.0.0`, I could remove that

Commit:
HollowFilesystemBlobStager: replaced File for Path

Commit:
Introduce FileManipulator and use Path for HollowFilesystemAnnouncer

Commit:
HollowFilesystemPublisher: use Path instead of File

Commit:
HollowFilesystemAnnouncementWatcher: use Path instead of File

Commit:
HollowFilesystemBlobRetriever: use Path instead of File

Commit:
Remove FileManipulator class

ghost commented:
I was behind on checking PRs. A pleasant surprise to see you did this! Reviewing now.

Commit:
Merge branch 'master' of https://github.com/Netflix/hollow into hollow3/switch-file-to-path

Commit:
changes based on feedback for HollowFilesystemAnnouncementWatcher and HollowFilesystemBlobRetriever

Commit:
changes based on feedback for HollowFilesystemAnnouncer

Commit:
comment deprecates with TODOs

Commit:
replace Path.get for resolve

rpalcolea commented:
Hi @toolbear,

Thanks for the feedback.

Made the changes:

* Replace `Paths.get` for `path.resolve()`
* Commented out `@Depredated` and added a TODO on each one
* Renamed all variables that were `Path` type and finished with File in the name to use Path instead.
* Changed all the `@since 3.0.0` to `@since 2.12.0`
* Improved the exception error messages

ghost commented:
Good stuff. :shipit:
